rdmo/projects/handlers/sync_utils.py

```diff
@@ -0,0 +1,83 @@
+import logging
+
+from rdmo.projects.handlers.utils import get_related_field_name_on_model_for_instance
+from rdmo.projects.models import Project
+
+logger = logging.getLogger(__name__)
+
+
+def sync_task_or_view_to_projects(instance):
+    """Ensure the instance is linked to exactly the correct set of projects."""
+    project_m2m_field = get_related_field_name_on_model_for_instance(Project, instance)
+    target_projects = Project.objects.filter_projects_for_task_or_view(instance)
+
+    current_projects = Project.objects.filter(**{project_m2m_field: instance})
+
+    to_remove = current_projects.exclude(pk__in=target_projects)
+    to_add = target_projects.exclude(pk__in=current_projects)
+    if not to_remove and not to_add:
+        logger.debug(
+            'Nothing to change for %s (%s) [sync_instance_to_projects]',
+            instance,
+            project_m2m_field
+        )
+        return
+
+    for project in to_remove:
+        logger.debug(
+            'Removing %s from %s(id=%s).%s [sync_instance_to_projects]',
+            instance,
+            project,
+            project.id,
+            project_m2m_field
+        )
+        getattr(project, project_m2m_field).remove(instance)
+
+    for project in to_add:
```

Comment:
`instance.projects.add(*to_add)`

Reply:
1 query per view instead of 1 per project :exploding_head:

Reply:
thanks, looks more optimized now!